src/lib.rs

```diff
@@ -154,14 +154,17 @@ where
 /// This is generic over the index. The type parameter for the direction is an implementation
 /// detail to ensure this is a type-level constant (instead of it being checked at runtime).
 #[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd)]
-pub struct Range<Idx, Direction> {
-    direction: Direction,
+pub struct Range<Idx, Dir>
+where
+    Dir: Direction<Idx>,
+{
+    direction: Dir,
     from: Idx,
     to: Idx,
 }
 
 /// Implement copy for Ranges with copyable indices
-impl<T: Copy, Direction: Copy> Copy for Range<T, Direction> {}
+impl<T: Copy, Dir: Copy + Direction<T>> Copy for Range<T, Dir> {}
```

Comment:
Oh, interesting: This seems to be the only case where we use `T` instead of `Idx`